datafusion/physical-expr/src/aggregate/moving_min_max.rs

```diff
@@ -0,0 +1,264 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// The implementation is taken from https://github.com/spebern/moving_min_max/blob/master/src/lib.rs.
+
+//! Keep track of the minimum or maximum value in a sliding window.
+//!
+//! `moving min max` provides one data structure for keeping track of the
+//! minimum value and one for keeping track of the maximum value in a sliding
+//! window.
+//!
+//! Each element is stored with the current min/max. One stack to push and another one for pop. If pop stack is empty,
+//! push to this stack all elements popped from first stack while updating their current min/max. Now pop from
+//! the second stack (MovingMin/Max struct works as a queue). To find the minimum element of the queue,
+//! look at the smallest/largest two elements of the individual stacks, then take the minimum of those two values.
+//!
+//! The complexity of the operations are
+//! - O(1) for getting the minimum/maximum
+//! - O(1) for push
+//! - amortized O(1) for pop
+
+/// let mut moving_min = MovingMin::<i32>::new();
+/// moving_min.push(2);
+/// moving_min.push(1);
+/// moving_min.push(3);
+///
+/// assert_eq!(moving_min.min(), Some(&1));
+/// assert_eq!(moving_min.pop(), Some(2));
+///
+/// assert_eq!(moving_min.min(), Some(&1));
+/// assert_eq!(moving_min.pop(), Some(1));
+///
+/// assert_eq!(moving_min.min(), Some(&3));
+/// assert_eq!(moving_min.pop(), Some(3));
+///
+/// assert_eq!(moving_min.min(), None);
+/// assert_eq!(moving_min.pop(), None);
```

Comment:
If you intend this to run as a doc comment, I think you need to put it in "```"

```suggestion
/// ```
/// let mut moving_min = MovingMin::<i32>::new();
/// moving_min.push(2);
/// moving_min.push(1);
/// moving_min.push(3);
///
/// assert_eq!(moving_min.min(), Some(&1));
/// assert_eq!(moving_min.pop(), Some(2));
///
/// assert_eq!(moving_min.min(), Some(&1));
/// assert_eq!(moving_min.pop(), Some(1));
///
/// assert_eq!(moving_min.min(), Some(&3));
/// assert_eq!(moving_min.pop(), Some(3));
///
/// assert_eq!(moving_min.min(), None);
/// assert_eq!(moving_min.pop(), None);
/// ```
```

Reply:
These tests still aren't running -- if we are going to have example code I think they should be in backticks so they are part of the doctests (and tested via `cargo test`)